### PR TITLE
Feature/lfa 138 determine strike off date

### PR DESCRIPTION
**Pull request by markpit**
Also made entities Immutable and renamed RestorationCalculatorResults to RestorationCalculatorResponse

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/restoration/corporatebody/CorporateBodyService.java [hunk 1]
```diff
@@ -2,14 +2,30 @@
...
     private CorporateBodyRepository corporateBodyRepository;
 
+    @Autowired
+    private TransactionRepository transactionRepository;
+
+    @Transactional(readOnly=true)
     public CorporateBody getCorporateBody(String incorporationNumber) {
-        return corporateBodyRepository.findByIncorporationNumber(incorporationNumber);
+        CorporateBody corporateBody = corporateBodyRepository.findByIncorporationNumber(incorporationNumber);
+        Optional<Date> strikeOffDate = transactionRepository.findLatestStrikeOffDate(corporateBody.getId());
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/restoration/corporatebody/CorporateBodyService.java]: Hibernate can't convert a timestamp to localDate so had to use Date instead and then do the conversion

File: src/main/java/uk/gov/companieshouse/restoration/transaction/Transaction.java [hunk 2]
```diff
@@ -30,5 +33,5 @@
     private Long transactionTypeId;
     
     @Column(name = "transaction_status_date")
-    private LocalDate transactionStatusDate;
+    private Date transactionStatusDate;
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/restoration/transaction/Transaction.java]: Hibernate throwing an error that it can't convert a timestamp to LocalDate so had to use a Date instead

File: src/main/java/uk/gov/companieshouse/restoration/corporatebody/CorporateBodyService.java (hunk 1 above)
Comment (jterrych) [reply src/main/java/uk/gov/companieshouse/restoration/corporatebody/CorporateBodyService.java]: Rather than have to use a Date, couldnt we use LocalDateTime?

Comment (jterrych) [reply src/main/java/uk/gov/companieshouse/restoration/corporatebody/CorporateBodyService.java]: Its also possible for a corporate body to be null here. should be be retrieving a strikeoffdate before the basic corporateBody has been validated? Maybe the CorporateBodyService should call the validationprocessor?

File: src/main/java/uk/gov/companieshouse/restoration/transaction/Transaction.java (hunk 2 above)
Comment (jterrych) [reply src/main/java/uk/gov/companieshouse/restoration/transaction/Transaction.java]: Same as above, can we avoid Date completely by using a LocalDateTime?

File: src/test/java/uk/gov/companieshouse/restoration/calculator/CorporateBodyServiceTest.java [hunk 3]
```diff
@@ -9,23 +9,35 @@
...
     static {
         GIRLS_SCHOOL_CORPORATE_BODY.setName(GIRLS_SCHOOL_NAME);
+        GIRLS_SCHOOL_CORPORATE_BODY.setId(GIRLS_SCHOOL_ID);
     }
 
     @Mock
     private CorporateBodyRepository mockCorporateBodyRepository;
 
+    @Mock
+    private TransactionRepository mockTransactionRepository;
```
Comment (jterrych) [inline src/test/java/uk/gov/companieshouse/restoration/calculator/CorporateBodyServiceTest.java]: Must add a test where corporateBody is null, not dissolved, not the correct status type and has been dissolved over 6 years

File: src/main/java/uk/gov/companieshouse/restoration/corporatebody/CorporateBody.java [hunk 4]
```diff
@@ -33,4 +37,7 @@
     @OneToOne
     @JoinColumn(name = "corporate_body_type_id")
     private CorporateBodyType corporateBodyType;
+
+    @Transient
```
Comment (mattch1) [inline src/main/java/uk/gov/companieshouse/restoration/corporatebody/CorporateBody.java]: Do we need this transient variable? Is it possible to just access the the strike off date at service level through the repository without causing any changes to be made to the db?

File: src/test/java/uk/gov/companieshouse/restoration/transaction/TransactionRepositoryTest.java [hunk 5]
```diff
@@ -23,8 +26,9 @@
     
     @Test
     public void canRetrieveSingleValue() {
-        LocalDate strikeOffDate = repo.findLatestStrikeOffDate(1L);
-                
-        assertThat(strikeOffDate, equalTo(LocalDate.of(2018, 01, 10)));
+        Optional<Date> strikeOffDate = repo.findLatestStrikeOffDate(1L);
```
Comment (mattch1) [inline src/test/java/uk/gov/companieshouse/restoration/transaction/TransactionRepositoryTest.java]: We should perhaps add a conditional to handle any empty optional that is returned here.

File: src/main/java/uk/gov/companieshouse/restoration/corporatebody/CorporateBody.java (hunk 4 above)
Comment (markpit) [reply src/main/java/uk/gov/companieshouse/restoration/corporatebody/CorporateBody.java]: Transient will be ignored by hibernate so DB won't get updated plus it's immutable. The alternative would be for the controller/calculator to call it - I thought it would be good for the corporate body to include it.

File: src/main/java/uk/gov/companieshouse/restoration/corporatebody/CorporateBodyService.java (hunk 1 above)
Comment (markpit) [reply src/main/java/uk/gov/companieshouse/restoration/corporatebody/CorporateBodyService.java]: I forgot about the null situation. We could add a dissolved check here but i guess we would be repeating the validation. So perhaps get initial data - then validate - then get extra data as and when we need it?

File: src/main/java/uk/gov/companieshouse/restoration/transaction/TransactionService.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+public class TransactionService {
+
+    @Autowired
+    private TransactionRepository transactionRepository;
+
+    @Transactional(readOnly=true)
+    public Optional<LocalDate> getStrikeOffDate(Long corporateBodyId) {
+        Optional<Timestamp> strikeOffDate = transactionRepository.findLatestStrikeOffDate(corporateBodyId);
+
+        return strikeOffDate
```
Comment (mattch1) [inline src/main/java/uk/gov/companieshouse/restoration/transaction/TransactionService.java]: Is it possible for optional to be empty here? If so do we need to handle it in some way?

Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/restoration/transaction/TransactionService.java]: Seen the "if present" code above - ignore these questions.
